### PR TITLE
cloud amazon ECS service modules

### PR DESCRIPTION
**Pull request by Java1Guy**
A new module for dealing with ECS services on AWS, including a fact-getting module.

### CODE REVIEW

**Comment by robynbergeron** [comment]
Hi @Java1Guy -- thanks for submitting this new module. I noticed that your module doesn't have an author: line in it with your name & github ID -- can you update that, so it adheres to the guidelines (http://docs.ansible.com/developing_modules.html#module-checklist)? Thanks! :)

**Comment by robynbergeron** [comment]
Almost forgot: when you're done, please add a comment with the text "ready_for_review" and we will continue with the review process.  Thanks!

**Comment by Java1Guy** [comment]
ready_for_review

**Comment by bennojoy** [comment]
@Java1Guy The documentation seems to be missing the parameters the module takes  and examples seems incomplete , can you please add some valid examples and also please check documentation via ansible-doc -M <path_to_module> module_name

**Comment by Java1Guy** [comment]
Documentation updated and checked with ansible-doc.

**Comment by Java1Guy** [comment]
ready_for_review

**Comment by gregdek** [comment]
Back into review. Ping @bennojoy since you've already reviewed once; would be great to get your +1 here. Also adding new process boilerplate.

# 

Thanks for submitting this new module to Ansible Extras! This module is now in community review, a process that is open to all Ansible users. In order for this module to be approved, it must gain the following votes:

“works_for_me”: If you have tested the module thoroughly, including testing of all of the module’s options, and if the module works for you, please add “works_for_me” in the comments.

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

“needs_revision”: If the module fails to work for you, or if it doesn’t meet guidelines, please add “needs_revision” in the comments with details about what needs to be fixed.

When a module has both “works_for_me” and “passes_guidelines” tags, we will promote the module for inclusion in Ansible Extras.  At this point, you will be expected to maintain the module by fixing bugs and evaluating pull requests in a timely manner. 

Thanks again for submitting your Ansible module!

**Comment by juliedavila** [comment]
shipit

**Comment by gregdek** [comment]
Thanks @defionscode for your review and thanks @Java1Guy for the new modules. Marking for inclusion.

**Comment by jefftougas** [comment]
@Java1Guy thanks for building this! This is very helpful to me on a project I'm working on. One quick question - any plans to add support for updating an existing ecs service? I'm not super well versed in python but based on my read through of your code it doesn't seem that updating an existing ecs service is supported - just creation and deletion. Correct me if I'm wrong here, please. Thanks!

**Comment by Java1Guy** [comment]
Jeff:
I'm glad to hear you will find this useful...  a few comments:
1. Despite being in the extras library, this work is funded by Ansible, Inc. - so the roadmap is up to them.
2. The "typical" way Ansible handles your request is that by characterizing the service and requesting state: "present", it would verify that the service exists AS YOU NOW REQUEST... so, ideally, it would detect any differences and perform essentially an update.  Your read of code is not far off - at this point it is mostly checking existence.  If there are particular parameters that make sense to change, include them in a comment here and I'll look at putting them in a future release!
Cheers,
Mark

**Comment by jefftougas** [comment]
Thanks for the quick response Mark and for the info.

The specific use case I'm looking to fulfill is updating an existing service with a new task definition, e.g. as part of a continuous delivery pipeline after the corresponding docker image has been built and is ready for deploy.

I'll keep an eye on this module, and in the meantime if I get a chance to add this functionality myself I'll be sure to submit a PR.

Amazon's documentation on updating an existing service: http://docs.aws.amazon.com/AmazonECS/latest/developerguide/update-service.html
Boto3's documentation on their ecs `update_service()` method: https://boto3.readthedocs.org/en/latest/reference/services/ecs.html#ECS.Client.update_service

To answer your question as far as particular parameters I'd be wanting to change as part of an update, it really is just the task definition (in the format `family:revision`).

Thanks!

File: cloud/amazon/ecs_service.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+    runningCount: The number of tasks in the cluster that are in the RUNNING state.
+    serviceArn: The Amazon Resource Name (ARN) that identifies the service. The ARN contains the arn:aws:ecs namespace, followed by the region of the service, the AWS account ID of the service owner, the service namespace, and then the service name. For example, arn:aws:ecs:region :012345678910 :service/my-service .
+    serviceName: A user-generated string used to identify the service
+    status: The valid values are ACTIVE, DRAINING, or INACTIVE.
+    taskDefinition: The ARN of a task definition to use for tasks in the service.
+# Delete service
+ansible_facts: When deleting a service, the values described above for the service prior to its deletion are returned.
+'''
+try:
+    import json
```
Comment (abadger) [inline cloud/amazon/ecs_service.py:117]: Note: No need to include json in this try: except.  json isn't part of boto.  and it's in the python stdlib.

**Comment by abadger** [comment]
One note, non-blocking.  I'll merge this now and if you want to change that you can submit it as a new pull request.  Thanks!

**Comment by juliedavila** [comment]
@abadger I think we should merge this, I can make the pr mod later. Would be nice if this made the 2.0 cut.

**Comment by abadger** [comment]
Merged.  Sorry for the long delay.  We've been releasing release candidates for 2.0.0 Since November 19th which has delayed getting new modules committed.  Now that this is merged it should be on track for 2.1 release (which shouldn't take anywhere near as long as the time between 1.9 and 2.0).

**Pull request merged**